sdlib/d/gc/tcache.d

```diff
@@ -74,7 +74,50 @@ public:
 		}
 
 		auto pd = getPageDescriptor(ptr);
-		pd.arena.free(emap, pd, ptr);
+		freeImpl(pd, ptr);
+	}
+
+	void destroy(void* ptr) {
+		if (ptr is null) {
+			return;
+		}
+
+		auto pd = getPageDescriptor(ptr);
+
+		auto e = pd.extent;
+		if (e is null) {
+			return;
+		}
+
+		if (pd.isSlab()) {
+			// Slab is not yet supported
+			return;
+		}
+
+		if (e.finalizer is null) {
+			return;
+		}
+
+		(cast(void function(void* body, size_t size)) e.finalizer)(
+			ptr, e.usedCapacity);
+
+		freeImpl(pd, ptr);
+	}
+
+	bool setFinalizer(void* ptr, void* finalizer) {
```

Comment:
You can't add a finalizer ex post facto because you don't know you have the room (you have it for large, but you can't know that from ptr).

You want to allocate with a finalizer.